### DISCUSSION TITLE
Block packages based on collaborator

### DISCUSSION THREAD

**Discussion by gr2m** [npm/feedback#629] [Registry]
Over the past years, I've seen actors in the npm eco system to cause harm intentionally to put pressure on their users. That is their choice, but it's became reason enough for me to make sure none of the packages that these users are collaborators on are trustworthy for me, and I don't want them to end up in my dependency trees.

Another type of users I'd like to block have repeatedly broken the semver contract in the past, or interpret it differently than I do. Yet another type are users who repeatedly published packages with bugs.

I don't judge any of these, but at least for some of my projects, I would very much like to block all packages that belong to a block list that I would like to define in my npm personal account or organization account.

If anyone has suggestions for workarounds, I'm all ears

**Comment by bnb** [comment]
I think defining "block" would be an important factor in how this could be correctly implemented, given the massively transitive nature of dependencies in the context of packages published to npm.

Do you not want to receive packages from that author? do you not want to include them in your own modules when you're publishing? do either of those include transitive dependencies or do they just include top-level dependencies? 

further, there's some pretty big remediation questions: how does the CLI handle when a downstream dependency adds a dependency from a blocked author to something you've built and already have in production? how does the CLI handle when you add something but there is a blocked author in a transitive dependency?

**Reply by gr2m** [reply]
> include transitive dependencies

yes. I want to block receiving their packages all across the dependency tree.

> further, there's some pretty big remediation questions: how does the CLI handle when a downstream dependency adds a dependency from a blocked author to something you've built and already have in production?

I use lock files in production, so it shouldn't be a problem.

> how does the CLI handle when you add something but there is a blocked author in a transitive dependency?

It should fail with an error and a URL in the documentation where workarounds can be described, such as per-project exceptions

**Reply by bnb** [reply]
> I use lock files in production, so it shouldn't be a problem.

in those situations are you okay with never updating that module again, presuming they never remove that transitive dependency?

> I use lock files in production, so it shouldn't be a problem.

how does this work for modules rather than applications, where you don't get the protections of a lockfile?

**Reply by ljharb** [reply]
I don't think it'd be possible for this to be a feature for packages - packages declare semver ranges, and "`^1.4.0 but also no packages published by X, Y, or Z" isn't a semver range. I think that, like lockfiles, this would be an application-level feature.

**Reply by gr2m** [reply]
> this would be an application-level feature

fair enough, it would be a hard fail for applications.

I would still like for my libraries to fail in CI when a blocked dependency enters the locked dependency tree, at least optionally.

**Reply by ljharb** [reply]
It seems like that could be achieved in a package's CI by a tool that listed all the publishers of your ideal dependency tree. It just so happens that I have a tool that will be able to do precisely that (likely for some of the exact reasons implied by the OP) published within the next week or two :-)

**Comment by coderbyheart** [comment]
One possible solution is to introduce a `denylist` in `package.json` which will make an `npm i` fail when trying to add a package that matches an entry in that list. It could also be solved for you / your organizations packages by a `preinstall` hook that implements this check.

**Reply by gr2m** [reply]
this might be a slightly different use case than mine. I want the check to not only apply to direct dependencies, but across all of the dependency tree. And not only when adding a package, but each time when running npm install / updating the lock file
